src/Chapter4.hs

```diff
@@ -673,6 +719,84 @@ Specifically,
  ❃ Implement the function to convert Tree to list
 -}
 
+data BinaryTree a
+    = NodeBT a (BinaryTree a) (BinaryTree a)
+    | EmptyBT
+    deriving (Show, Read, Eq)
+
+instance Functor BinaryTree where
+    fmap _ EmptyBT = EmptyBT
+    fmap f (NodeBT x left right) = NodeBT (f x) (fmap f left) (fmap f right)
+
+
+-- | Switch left and right subtree of a tree recursively
+reverseTree :: BinaryTree a -> BinaryTree a
+reverseTree EmptyBT = EmptyBT
+reverseTree (NodeBT x left right) = NodeBT x (reverseTree right) (reverseTree left)
+
+
+data TraversalType
+    = PostOrder  -- ^ Starts from left subtree then right subtree then root
+    | PreOrder   -- ^ Starts from Root then left subtree and right subtree
+    | InOrder    -- ^ Starts from left subtree then root then right subtree
+
+
+preOrder :: BinaryTree a -> [a]
+preOrder EmptyBT = []
+preOrder (NodeBT x left right) = x : preOrder left ++ preOrder right
+
+postOrder :: BinaryTree a -> [a]
+postOrder EmptyBT = []
+postOrder (NodeBT x left right) = postOrder left ++ postOrder right ++ [x]
+
+inOrder :: BinaryTree a -> [a]
+inOrder EmptyBT = []
+inOrder (NodeBT x left right) = inOrder left ++ [x] ++ inOrder right
```

Comment:
Minor performance optimization, but here you can use `:` as well instead of `[x] ++`:

```suggestion
inOrder (NodeBT x left right) = inOrder left ++ (x : inOrder right)
```